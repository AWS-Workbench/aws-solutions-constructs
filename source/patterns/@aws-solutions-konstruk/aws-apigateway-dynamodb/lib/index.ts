/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as api from '@aws-cdk/aws-apigateway';
import * as iam from '@aws-cdk/aws-iam';
import * as defaults from '@aws-solutions-konstruk/core';
import { Construct } from '@aws-cdk/core';
import * as dynamodb from '@aws-cdk/aws-dynamodb';
import { overrideProps } from '@aws-solutions-konstruk/core';

/**
 * @summary The properties for the ApiGatewayToDynamoDB class.
 */
export interface ApiGatewayToDynamoDBProps {
  /**
   * Optional user provided props to override the default props
   *
   * @default - Default props are used
   */
  readonly dynamoTableProps?: dynamodb.TableProps,
  /**
   * Optional user-provided props to override the default props for the API Gateway.
   *
   * @default - Default properties are used.
   */
  readonly apiGatewayProps?: api.RestApiProps | any,

  /**
   * Whether to deploy API Gateway Method for Create operation on Dynamodb DB table.
   *
   * @default - false
   */
  readonly allowCreateOperation?: boolean,

  /**
   * API Gateway Request template for Create method, required if allowCreateOperation set to true
   *
   * @default - None
   */
  readonly createRequestTemplate?: string,

  /**
   * Whether to deploy API Gateway Method for Read operation on Dynamodb DB table.
   *
   * @default - true
   */
  readonly allowReadOperation?: boolean,

  /**
   * Whether to deploy API Gateway Method for Update operation on Dynamodb DB table.
   *
   * @default - false
   */
  readonly allowUpdateOperation?: boolean,

  /**
   * API Gateway Request template for Update method, required if allowUpdateOperation set to true
   *
   * @default - None
   */
  readonly updateRequestTemplate?: string,

  /**
   * Whether to deploy API Gateway Method for Delete operation on Dynamodb DB table.
   *
   * @default - false
   */
  readonly allowDeleteOperation?: boolean
}

/**
 * @summary The ApiGatewayToDynamoDB class.
 */
export class ApiGatewayToDynamoDB extends Construct {
  private table: dynamodb.Table;
  private apiGatewayRole: iam.Role;
  private apiGateway: api.RestApi;

  /**
   * @summary Constructs a new instance of the ApiGatewayToDynamoDB class.
   * @param {cdk.App} scope - represents the scope for all the resources.
   * @param {string} id - this is a a scope-unique id.
   * @param {CloudFrontToApiGatewayToLambdaProps} props - user provided props for the construct.
   * @since 0.8.0
   * @access public
   */
  constructor(scope: Construct, id: string, props: ApiGatewayToDynamoDBProps) {
    super(scope, id);
    let partitionKeyName: string;

    // Set the default props for DynamoDB table
    if (props.dynamoTableProps) {
      const dynamoTableProps: dynamodb.TableProps = overrideProps(defaults.DefaultTableProps, props.dynamoTableProps);
      partitionKeyName = dynamoTableProps.partitionKey.name;
      this.table = new dynamodb.Table(this, 'DynamoTable', dynamoTableProps);
    } else {
      partitionKeyName = defaults.DefaultTableProps.partitionKey.name;
      this.table = new dynamodb.Table(this, 'DynamoTable', defaults.DefaultTableProps);
    }

    // Setup the API Gateway
    this.apiGateway = defaults.GlobalRestApi(this);

    // Setup the API Gateway role
    this.apiGatewayRole = new iam.Role(this, 'api-gateway-role', {
      assumedBy: new iam.ServicePrincipal('apigateway.amazonaws.com')
    });

    // Setup the API Gateway Resource
    const apiGatewayResource: api.Resource = this.apiGateway.root.addResource("{" + partitionKeyName + "}");

    // Setup API Gateway Method
    // Create
    if (props.allowCreateOperation && props.allowCreateOperation === true && props.createRequestTemplate) {
      const createRequestTemplate = props.createRequestTemplate.replace("${Table}", this.table.tableName);
      this.addActiontoPlicy("dynamodb:PutItem");
      this.addMethod(this.apiGateway.root, createRequestTemplate, "PutItem", "POST");
    }
    // Read
    if (!props.allowReadOperation || props.allowReadOperation === true) {
      const getRequestTemplate = "{\r\n\"TableName\": \"" + this.table.tableName + "\",\r\n \"KeyConditionExpression\": \"" + partitionKeyName + " = :v1\",\r\n    \"ExpressionAttributeValues\": {\r\n        \":v1\": {\r\n            \"S\": \"$input.params('" + partitionKeyName + "')\"\r\n        }\r\n    }\r\n}";
      this.addActiontoPlicy("dynamodb:Query");
      this.addMethod(apiGatewayResource, getRequestTemplate, "Query", "GET");
    }
    // Update
    if (props.allowUpdateOperation && props.allowUpdateOperation === true && props.updateRequestTemplate) {
      const updateRequestTemplate = props.updateRequestTemplate.replace("${Table}", this.table.tableName);
      this.addActiontoPlicy("dynamodb:UpdateItem");
      this.addMethod(apiGatewayResource, updateRequestTemplate, "UpdateItem", "PUT");
    }
    // Delete
    if (props.allowDeleteOperation && props.allowDeleteOperation === true) {
      const deleteRequestTemplate = "{\r\n  \"TableName\": \"" + this.table.tableName + "\",\r\n  \"Key\": {\r\n    \"" + partitionKeyName + "\": {\r\n      \"S\": \"$input.params('" + partitionKeyName + "')\"\r\n    }\r\n  },\r\n  \"ConditionExpression\": \"attribute_not_exists(Replies)\",\r\n  \"ReturnValues\": \"ALL_OLD\"\r\n}";
      this.addActiontoPlicy("dynamodb:DeleteItem");
      this.addMethod(apiGatewayResource, deleteRequestTemplate, "DeleteItem", "DELETE");
    }
  }

  private addActiontoPlicy(action: string) {
    this.apiGatewayRole.addToPolicy(new iam.PolicyStatement({
      resources: [
        this.table.tableArn
      ],
      actions: [ `${action}` ]
    }));
  }

  private addMethod(apiResource: api.IResource, requestTemplate: string, dynamodbAction: string, apiMethod: string) {
    // Setup the API Gateway Integration
    const apiGatewayIntegration = new api.AwsIntegration({
      service: "dynamodb",
      action: dynamodbAction,
      integrationHttpMethod: "POST",
      options: {
        passthroughBehavior: api.PassthroughBehavior.NEVER,
        credentialsRole: this.apiGatewayRole,
        requestParameters: {
            "integration.request.header.Content-Type": "'application/json'"
        },
        requestTemplates: {
            "application/json": requestTemplate
        },
        integrationResponses: [
          {
              statusCode: "200"
          },
          {
              statusCode: "500",
              responseTemplates: {
                  "text/html": "Error"
              },
              selectionPattern: "500"
          }
        ]
      }
    });

    // Setup the API Gateway method(s)
    apiResource.addMethod(apiMethod, apiGatewayIntegration, {
        methodResponses: [
            {
                statusCode: "200",
                responseParameters: {
                    "method.response.header.Content-Type": true
                }
            },
            {
                statusCode: "500",
                responseParameters: {
                    "method.response.header.Content-Type": true
                },
            }
        ]
    });
  }

  /**
   * @summary Returns an instance of the api.RestApi created by the construct.
   * @returns {api.RestApi} Instance of the RestApi created by the construct.
   * @since 0.8.0
   * @access public
   */
  public restApi(): api.RestApi {
    return this.apiGateway;
  }

  /**
   * @summary Returns an instance of dynamodb.Table created by the construct.
   * @returns {dynamodb.Table} Instance of dynamodb.Table created by the construct
   * @since 0.8.0
   * @access public
   */
  public dynamoTable(): dynamodb.Table {
    return this.table;
  }
}