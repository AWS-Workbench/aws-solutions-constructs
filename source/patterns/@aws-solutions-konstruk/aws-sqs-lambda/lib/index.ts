/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as sqs from '@aws-cdk/aws-sqs';
import * as lambda from '@aws-cdk/aws-lambda';
import * as kms from '@aws-cdk/aws-kms';
import * as defaults from '@aws-solutions-konstruk/core';
import { Construct } from '@aws-cdk/core';
import { SqsEventSource } from '@aws-cdk/aws-lambda-event-sources';

/**
 * @summary The properties for the SqsToLambda class.
 */
export interface SqsToLambdaProps {
    /**
     * Whether to create a new Lambda function or use an existing Lambda function.
     * If set to false, you must provide an existing function for the `existingLambdaObj` property.
     *
     * @default - true
     */
    readonly deployLambda: boolean,
    /**
     * Existing instance of Lambda Function object.
     * If `deploy` is set to false only then this property is required
     *
     * @default - None
     */
    readonly existingLambdaObj?: lambda.Function,
    /**
     * Optional user provided properties to override the default properties for the Lambda function.
     * If `deploy` is set to true only then this property is required.
     *
     * @default - Default properties are used.
     */
    readonly lambdaFunctionProps?: lambda.FunctionProps | any
    /**
     * Optional user provided properties
     *
     * @default - Default props are used
     */
    readonly queueProps?: sqs.QueueProps | any
    /**
     * Optional user provided props to override the default props for the KMS.
     *
     * @default - Default props are used
     */
    readonly encryptionKeyProps?: kms.KeyProps | any
    /**
     * Whether to deploy a secondary queue to be used as a dead letter queue.
     *
     * @default - required field.
     */
    readonly deployDeadLetterQueue: boolean,
    /**
     * The number of times a message can be unsuccesfully dequeued before being moved to the dead-letter queue.
     *
     * @default - required field.
     */
    readonly maxReceiveCount: number
}

/**
 * @summary The SqsToLambda class.
 */
export class SqsToLambda extends Construct {
    // Private variables
    private queue: sqs.Queue;
    private fn: lambda.Function;
    private encryptionKey: kms.Key;

    /**
     * @summary Constructs a new instance of the SqsToLambda class.
     * @param {cdk.App} scope - represents the scope for all the resources.
     * @param {string} id - this is a a scope-unique id.
     * @param {CloudFrontToApiGatewayToLambdaProps} props - user provided props for the construct.
     * @since 0.8.0
     * @access public
     */
    constructor(scope: Construct, id: string, props: SqsToLambdaProps) {
        super(scope, id);

        // Setup the encryption key
        this.encryptionKey = defaults.buildEncryptionKey(scope, props.encryptionKeyProps);

        // Setup the Lambda function
        this.fn = defaults.buildLambdaFunction(scope, {
            deployLambda: props.deployLambda,
            existingLambdaObj: props.existingLambdaObj,
            lambdaFunctionProps: props.lambdaFunctionProps
        });

        // Setup the dead letter queue, if applicable
        let dlqi: sqs.DeadLetterQueue | undefined;
        if (props.deployDeadLetterQueue) {
            const dlq: sqs.Queue = defaults.buildQueue(scope, 'deadLetterQueue', {
                encryptionKey: this.encryptionKey,
                queueProps: props.queueProps
            });
            dlqi = defaults.buildDeadLetterQueue({
                deadLetterQueue: dlq,
                maxReceiveCount: props.maxReceiveCount
            });
        }

        // Setup the queue
        this.queue = defaults.buildQueue(scope, 'queue', {
            encryptionKey: this.encryptionKey,
            queueProps: props.queueProps,
            deadLetterQueue: dlqi
        });

        // Setup the event source mapping
        this.fn.addEventSource(new SqsEventSource(this.queue));
    }

    /**
     * @summary Returns an instance of the lambda.Function created by the construct.
     * @returns {lambda.Function} Instance of the Function created by the construct.
     * @since 0.8.0
     * @access public
     */
    public lambdaFunction(): lambda.Function {
        return this.fn;
    }

    /**
     * @summary Returns an instance of the sqs.Queue created by the construct.
     * @returns {sqs.Queue} Instance of the Queue created by the construct.
     * @since 0.8.0
     * @access public
     */
    public sqsQueue(): sqs.Queue {
        return this.queue;
    }
}