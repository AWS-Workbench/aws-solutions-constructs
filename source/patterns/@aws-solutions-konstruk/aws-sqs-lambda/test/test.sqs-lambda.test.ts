/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import { Stack } from "@aws-cdk/core";
import { SqsToLambda, SqsToLambdaProps } from "../lib";
import * as lambda from '@aws-cdk/aws-lambda';
import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';

// --------------------------------------------------------------
// Pattern deployment w/ new Lambda function and
// default properties
// --------------------------------------------------------------
test('Pattern deployment w/ new Lambda function and default props', () => {
    // Initial Setup
    const stack = new Stack();
    const props: SqsToLambdaProps = {
        deployLambda: true,
        lambdaFunctionProps: {
            runtime: lambda.Runtime.NODEJS_10_X,
            handler: 'index.handler',
            code: lambda.Code.asset(`${__dirname}/lambda`)
        },
        deployDeadLetterQueue: true,
        maxReceiveCount: 15,
        queueProps: {},
        encryptionKeyProps: {},
    };
    new SqsToLambda(stack, 'test-sqs-lambda', props);
    // Assertion 1
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

// --------------------------------------------------------------
// Pattern deployment w/ new Lambda function and
// overridden properties
// --------------------------------------------------------------
test('Pattern deployment w/ new Lambda function and overridden props', () => {
    // Initial Setup
    const stack = new Stack();
    const props: SqsToLambdaProps = {
        deployLambda: true,
        lambdaFunctionProps: {
            runtime: lambda.Runtime.NODEJS_10_X,
            handler: 'index.handler',
            code: lambda.Code.asset(`${__dirname}/lambda`),
            environment: {
                OVERRIDE: "TRUE"
            }
        },
        queueProps: {
            fifo: true
        },
        encryptionKeyProps: {},
        deployDeadLetterQueue: false,
        maxReceiveCount: 0
    };
    const app = new SqsToLambda(stack, 'test-sqs-lambda', props);
    // Assertion 1
    expect(app.lambdaFunction()).toHaveProperty('environment.OVERRIDE', 'TRUE');
    // Assertion 2
    expect(app.sqsQueue()).toHaveProperty('fifo', true);
});

// --------------------------------------------------------------
// Pattern Deployment w/ Existing Lambda function
// --------------------------------------------------------------
test('Pattern deployment w/ Existing Lambda Function', () => {
    // Initial Setup
    const stack = new Stack();
    const fn = new lambda.Function(stack, 'ExistingLambdaFunction', {
        runtime: lambda.Runtime.NODEJS_10_X,
        handler: 'index.handler',
        code: lambda.Code.asset(`${__dirname}/lambda`)
    });
    const props: SqsToLambdaProps = {
        deployLambda: false,
        existingLambdaObj: fn,
        deployDeadLetterQueue: false,
        encryptionKeyProps: {},
        maxReceiveCount: 0,
        queueProps: {}
    };
    new SqsToLambda(stack, 'test-apigateway-lambda', props);
    // Assertion 1
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

// --------------------------------------------------------------
// Test the getter methods
// --------------------------------------------------------------
test('Test getter methods', () => {
    // Initial Setup
    const stack = new Stack();
    const props: SqsToLambdaProps = {
        deployLambda: true,
        lambdaFunctionProps: {
            runtime: lambda.Runtime.NODEJS_10_X,
            handler: 'index.handler',
            code: lambda.Code.asset(`${__dirname}/lambda`)
        },
        deployDeadLetterQueue: false,
        encryptionKeyProps: {},
        maxReceiveCount: 0,
        queueProps: {}
    };
    const app = new SqsToLambda(stack, 'test-apigateway-lambda', props);
    // Assertion 1
    expect(app.lambdaFunction()).toBeDefined();
    // Assertion 2
    expect(app.sqsQueue()).toBeDefined();
});

// --------------------------------------------------------------
// Test error handling for existing Lambda function
// --------------------------------------------------------------
test('Test error handling for existing Lambda function', () => {
    // Initial Setup
    const stack = new Stack();
    const props: SqsToLambdaProps = {
        deployLambda: false,
        existingLambdaObj: undefined,
        deployDeadLetterQueue: false,
        encryptionKeyProps: {},
        maxReceiveCount: 0,
        queueProps: {}
    };
    // Assertion 1
    expect(() => {
        new SqsToLambda(stack, 'test-sqs-lambda', props);
    }).toThrowError();
});

// --------------------------------------------------------------
// Test error handling for new Lambda function
// w/o required properties
// --------------------------------------------------------------
test('Test error handling for new Lambda function w/o required properties', () => {
    // Initial Setup
    const stack = new Stack();
    const props: SqsToLambdaProps = {
        deployLambda: true,
        deployDeadLetterQueue: false,
        encryptionKeyProps: {},
        maxReceiveCount: 0,
        queueProps: {}
    };
    // Assertion 1
    expect(() => {
        new SqsToLambda(stack, 'test-sqs-lambda', props);
    }).toThrowError();
});