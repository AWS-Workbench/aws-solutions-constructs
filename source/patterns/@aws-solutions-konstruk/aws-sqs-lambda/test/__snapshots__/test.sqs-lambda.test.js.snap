// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pattern deployment w/ Existing Lambda Function 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420ArtifactHashA71E92AD": Object {
      "Description": "Artifact hash for asset \\"8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420\\"",
      "Type": "String",
    },
    "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3Bucket749AC458": Object {
      "Description": "S3 bucket for asset \\"8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420\\"",
      "Type": "String",
    },
    "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3VersionKeyFF5CC16D": Object {
      "Description": "S3 key for asset version \\"8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ExistingLambdaFunctionF606C520": Object {
      "DependsOn": Array [
        "ExistingLambdaFunctionServiceRoleDefaultPolicy2431D213",
        "ExistingLambdaFunctionServiceRole7CC6DE65",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3Bucket749AC458",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3VersionKeyFF5CC16D",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3VersionKeyFF5CC16D",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ExistingLambdaFunctionServiceRole7CC6DE65",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ExistingLambdaFunctionServiceRole7CC6DE65": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExistingLambdaFunctionServiceRoleDefaultPolicy2431D213": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testapigatewaylambdaqueue1FFAE03C",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testapigatewaylambdaEncryptionKey06FC467F",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ExistingLambdaFunctionServiceRoleDefaultPolicy2431D213",
        "Roles": Array [
          Object {
            "Ref": "ExistingLambdaFunctionServiceRole7CC6DE65",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ExistingLambdaFunctionSqsEventSourcetestapigatewaylambdaqueueFD30FF33BE4927D4": Object {
      "Properties": Object {
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "testapigatewaylambdaqueue1FFAE03C",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "ExistingLambdaFunctionF606C520",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "testapigatewaylambdaEncryptionKey06FC467F": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "ExistingLambdaFunctionServiceRole7CC6DE65",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "testapigatewaylambdaqueue1FFAE03C": Object {
      "Properties": Object {
        "KmsMasterKeyId": Object {
          "Fn::GetAtt": Array [
            "testapigatewaylambdaEncryptionKey06FC467F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SQS::Queue",
    },
  },
}
`;

exports[`Pattern deployment w/ new Lambda function and default props 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420ArtifactHashA71E92AD": Object {
      "Description": "Artifact hash for asset \\"8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420\\"",
      "Type": "String",
    },
    "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3Bucket749AC458": Object {
      "Description": "S3 bucket for asset \\"8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420\\"",
      "Type": "String",
    },
    "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3VersionKeyFF5CC16D": Object {
      "Description": "S3 key for asset version \\"8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "testsqslambdaEncryptionKey317A2F03": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "testsqslambdaLambdaFunctionServiceRoleF623B438",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "testsqslambdaLambdaFunction58720146": Object {
      "DependsOn": Array [
        "testsqslambdaLambdaFunctionServiceRoleDefaultPolicy380B065C",
        "testsqslambdaLambdaFunctionServiceRoleF623B438",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3Bucket749AC458",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3VersionKeyFF5CC16D",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8efd3dd9643a4d64a128ad582cab718a1e464bcc719bbbcf0e7b0481188a0420S3VersionKeyFF5CC16D",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testsqslambdaLambdaFunctionServiceRoleF623B438",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testsqslambdaLambdaFunctionServiceRoleDefaultPolicy380B065C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testsqslambdaqueue601203B8",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testsqslambdaEncryptionKey317A2F03",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testsqslambdaLambdaFunctionServiceRoleDefaultPolicy380B065C",
        "Roles": Array [
          Object {
            "Ref": "testsqslambdaLambdaFunctionServiceRoleF623B438",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testsqslambdaLambdaFunctionServiceRoleF623B438": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:aws:logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testsqslambdaLambdaFunctionSqsEventSourcetestsqslambdaqueue583E2E6CB891E0FC": Object {
      "Properties": Object {
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "testsqslambdaqueue601203B8",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "testsqslambdaLambdaFunction58720146",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "testsqslambdadeadLetterQueue85BDB0A3": Object {
      "Properties": Object {
        "KmsMasterKeyId": Object {
          "Fn::GetAtt": Array [
            "testsqslambdaEncryptionKey317A2F03",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SQS::Queue",
    },
    "testsqslambdaqueue601203B8": Object {
      "Properties": Object {
        "KmsMasterKeyId": Object {
          "Fn::GetAtt": Array [
            "testsqslambdaEncryptionKey317A2F03",
            "Arn",
          ],
        },
        "RedrivePolicy": Object {
          "deadLetterTargetArn": Object {
            "Fn::GetAtt": Array [
              "testsqslambdadeadLetterQueue85BDB0A3",
              "Arn",
            ],
          },
          "maxReceiveCount": 15,
        },
      },
      "Type": "AWS::SQS::Queue",
    },
  },
}
`;
